[Channel] re-implement using a state machine

twittemb opened the pull request:
Hi

This PR suggests a new implementation for `AsyncChannel`. The idea was to:

- streamline the implementation using the same paradigm (state machine/storage) as the recently remade operators (merge, zip, ...)
- share some code between `AsyncChannel` and `AsyncThrowingChannel` since the state machine is the same
- eventually improve the performance (I've seen a 20% increase in throughput)

Some remarks:
- I've added a throughput measurement
- I think there is a bug in the old/current implementation of `AsyncChannel` because when I tried to run  it with the new throughput measurement I had a crash from time to time
- with this new implementation we could imagine to split the producer from the consumer (same as what @FranzBusch pitched here https://forums.swift.org/t/pitch-convenience-async-throwing-stream-makestream-methods/61030 for AsyncStream). We just have to reference the same storage from an hypothetical `AsyncChannel.Continuation` and the `AsyncChannel` it self.

@phausler @FranzBusch what do you think ?

phausler commented:
The perf sounds fantastic! I will make sure to go over this with a fine tooth comb Monday.

twittemb commented:
For the record, I've used ad-hoc properties (hidden behind `#if DEBUG` instructions) so we can test the suspend/resume mechanism more precisely, particularly for task cancellation. I know this is not an ideal solution but I can't see something better right now for this use case.

I guess this is a part of the resolution of https://github.com/apple/swift-async-algorithms/issues/148

FranzBusch commented:
Currently on vacation but will try to give this a look soon! I think @phausler has some ideas around testing that doesn’t involve scattering debug only code.

eaigner commented:
Just FYI, your bug is probably that you increment `generation += 1`. In swift this does not automatically overflow, but will crash if you reach `Int.max`.

twittemb commented:
> Just FYI, your bug is probably that you increment `generation += 1`. In swift this does not automatically overflow, but will crash if you reach `Int.max`.

Hi, the bug

eaigner commented:
Also the implementation seems overcomplicated to me. I have a buffered channel here `https://github.com/eaigner/AsyncChannel` which uses only 130 LOC. Granted, no cancellation, but still.

phausler commented:
Cancellation complicates things a decent amount; accounting for all the states and their cancellation counterparts ends up being roughly similar. The advantage with state machines is that the states are well known about which transition it can be and explicitly calls those out - so for subtile state manipulation that can cross boundaries of tasks it is useful to have a bit more # of lines of code (without much perf impact) to build known state transition clarity.

FranzBusch commented:
Agree with @phausler here LoC is not a good measurement here. `AsyncChannel` has some inherent complexities and state machines just surface the various transition edges way better.

FranzBusch commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,283 @@
+    case resumeProducerAndConsumer(UnsafeContinuation<Void, Never>?, Element?)
+  }
+
+  mutating func newConsumer(
```
Would prefer if we rename this to `next`

twittemb replied:
I've renames pretty much all the functions, please advise.

FranzBusch commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,152 @@
+    }
+  }
+
+  func fail(error: Error) where Failure == Error {
```
Why do we need this generic constraint here?

twittemb replied:
That is a constraint that already exists in the current implementation of `AsyncThrowingChannel` and that I also applied on the storage. I guess as the storage is an internal implementation would could remove that constraint indeed.

FranzBusch replied:
Yeah I think there is no value on having this constraint here

twittemb replied:
I've removed the constraint.

FranzBusch commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  func next() async throws -> Element? {
+    guard !Task.isCancelled else { return nil }
```
I wouldn't do both a `Task.isCancelled` and a cancellationHandler. IMO we can just remove this check

twittemb replied:
I've removed this line.

FranzBusch commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,152 @@
+    guard !Task.isCancelled else { return nil }
+
+    let consumerId = self.generateId()
+    let wasCancelled = ManagedCriticalState(false)
+
+    return try await withTaskCancellationHandler {
+      return try await withUnsafeThrowingContinuation { (continuation: UnsafeContinuation<Element?, any Error>) in
```
Same comments as above. I don't think we should track `wasCancelled` here. Furthermore, we should only conditionally create a continuation.

twittemb replied:
See my comment about tracking the already cancelled send/next thanks to `Set`s in the channeling state.

FranzBusch reviewed (commented):
Left some comments here! In general, I like the switch to a state machine it makes it very clear what is going on

twittemb commented:
> Left some comments here! In general, I like the switch to a state machine it makes it very clear what is going on

Thanks for your review … I‘ll check it ASAP

twittemb commented on Sources/AsyncAlgorithms/Channels/AsyncChannel.swift:
```diff
@@ -0,0 +1,70 @@
+
+  let storage: ChannelStorage<Element, Never>
+
+#if DEBUG
+  var onSendSuspended: (() -> Void)? {
+    didSet {
+      self.storage.onSendSuspended = onSendSuspended
+    }
+  }
+  var onNextSuspended: (() -> Void)? {
+    didSet {
+      self.storage.onNextSuspended = onNextSuspended
+    }
+  }
+#endif
```
You seemed to say that @phausler might have a solution for that ? The goal of these ad-hoc properties is the test the suspend/resume mechanism … what alternative do we have ?

twittemb replied:
This is the only thing I've not addressed yet, because I don't know how to keep the same code coverage without it :-(

twittemb replied:
@phausler do you have an opinion on that ?

phausler replied:
I am in agreement that debug builds should be the same code as release builds being built. Adding in these feels like an anti-pattern and likely with some thought/time dedicated to it we could probably use the custom scheduler stuff in the validation diagrams to ensure these invariants.

phausler replied:
shorter way of saying it; remove em for now and we can fix that coverage/testing later.

twittemb replied:
Sure, I'll remove that and push something tomorrow. Thanks.

twittemb replied:
It's been pushed.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,152 @@
+    await withTaskCancellationHandler {
+      await withUnsafeContinuation { (continuation: UnsafeContinuation<Void, Never>) in
+        self.stateMachine.withCriticalRegion { stateMachine in
+          let wasCancelled = wasCancelled.withCriticalRegion { $0 }
+          guard !wasCancelled else {
+            continuation.resume()
+            return
+          }
```
Yep that makes sense, I’ll try to push the logic into the state machine

twittemb replied:
the logic has been pushed in the state machine. The idea was to keep a list of cancelled producers/consumers so we can know of the cancellation has happened before the suspension.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,152 @@
+    let producerId = self.generateId()
+    let wasCancelled = ManagedCriticalState(false)
+
+    await withTaskCancellationHandler {
+      await withUnsafeContinuation { (continuation: UnsafeContinuation<Void, Never>) in
```
About that, one of my previous implementation of the buffer algorithm (currently in draft PR) involved checking the need to suspend before actually suspending. From a discussion on the forum (https://forums.swift.org/t/clarification-needed-on-unsafecontinuation-documentation/57803/9) it might be counter productive to do that. It seems the caller is not suspended if the continuation is resumed synchronously. 

Checking if we should suspend and then creating the continuation or not might be an overhead VS creating the continuation anyway.

I don’t have the expertise to decide that, perhaps @phausler could jump in?

FranzBusch replied:
The problem is not soo much about the caller being suspended or not but the actual work of setting up a cancellation handler and creating an unsafe continuation. While the latter is relatively cheap since it is unsafe. The former actually does have a noticeable overhead. So actually checking if we need to suspend or not should bring in performance gains

twittemb replied:
I've added a step where we make an attempt to `send` or to `next`, if we can (we are already waiting for producers or consumers) then we confirm the operation and we don't suspend if we can't we confirm the suspension.

twittemb commented on Sources/AsyncAlgorithms/Channels/AsyncChannel.swift:
```diff
@@ -0,0 +1,70 @@
+    let storage: ChannelStorage<Element, Never>
+
+    public mutating func next() async -> Element? {
+      try? await self.storage.next()
```
nope, can't throw.
fixed.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,283 @@
+    case resumeProducerAndConsumer(continuation: UnsafeContinuation<Element?, any Error>?)
+  }
+
+  mutating func newProducer(
```
I've renamed pretty much all the functions. please advise.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,283 @@
+  private enum State {
+    case idle
+    case waitingForProducer(suspendedConsumers: OrderedSet<SuspendedConsumer>)
+    case waitingForConsumer(suspendedProducers: OrderedSet<SuspendedProducer>)
```
I've merged idle/waitingForProducer/waitingForConsumer into `channeling` and I use the `emptiness` of the sets to infer the state.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,283 @@
+    case waitingForProducer(suspendedConsumers: OrderedSet<SuspendedConsumer>)
+    case waitingForConsumer(suspendedProducers: OrderedSet<SuspendedProducer>)
+    case terminated(Termination)
+    case modifying
```
I've removed that state.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,283 @@
+    case resumeConsumers(continuations: [UnsafeContinuation<Element?, any Error>?])
+  }
+
+  mutating func finish() -> FinishAction {
```
I've merged the two functions into `finish(error: Failure?)`

twittemb commented:
> Left some comments here! In general, I like the switch to a state machine it makes it very clear what is going on

Hi @FranzBusch I've pushed a version with pretty much all the comments addressed.

FranzBusch commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+      case .terminated(.finished):
+        return .resumeConsumer(element: nil)
+
+      case .terminated(.failed(let error)):
```
I don't think that this is aligned with the rule that `next()` should throw an error once and after that it should return `nil`. This code right now throws the same error over and over again. Implementing the expected behaviour is going to get interesting since it means that we have to keep track of each iterator and what they saw. Maybe we can achieve this by creating a single ID for an iterator that we keep inside the iterator and use for every next() call on the state machine instead of generating an ID per next call.

twittemb replied:
It is not exactly what happens. When the state is .finished(.failure) and there are some suspended consumers then they all receive the failure and the state is set to  .terminated(.finished). The subsequent call to next will receive “nil”.

twittemb replied:
@FranzBusch, @phausler before making any changes I'd like to be sure of the expected behaviour. I think that what I have now is aligned with the previous implementation of `AsyncChannel`.

FranzBusch replied:
Maybe I am misreading the code here as well. If the behaviour is the same then it should be good!

twittemb replied:
this is the unit test that checks that:

```swift
  func test_asyncThrowingChannel_resumes_producers_and_discards_additional_elements_when_fail_is_called() async throws {
 // Given: an AsyncThrowingChannel
 let sut = AsyncThrowingChannel<Int, Error>()

 // Given: 2 suspended send operations
 let task1 = Task {
   await sut.send(1)
 }

 let task2 = Task {
   await sut.send(2)
 }

 // When: failing the channel
 sut.fail(Failure())

 // Then: the send operations are resumed
 _ = await (task1.value, task2.value)

 // When: sending an extra value
 await sut.send(3)

 // Then: the send operation is resumed
 // Then: the iteration is resumed with a failure
 var collected = [Int]()
 do {
   for try await element in sut {
     collected.append(element)
   }
 } catch {
   XCTAssertTrue(collected.isEmpty)
   XCTAssertEqual(error as? Failure, Failure())
 }

 // When: requesting a next value
 var iterator = sut.makeAsyncIterator()
 let pastFailure = try await iterator.next()

 // Then: the past failure is nil
 XCTAssertNil(pastFailure)
}
```

FranzBusch replied:
Thanks for posting this. Do we also have a test that sets up an iterator `next()` call before and then calls `sut.fail()`?

twittemb replied:
yep

```swift
func test_asyncThrowingChannel_resumes_consumers_when_fail_is_called() async throws {
  // Given: an AsyncThrowingChannel
  let sut = AsyncThrowingChannel<Int, Error>()

  // Given: 2 suspended iterations
  let task1 = Task<Int?, Error> {
    var iterator = sut.makeAsyncIterator()
    return try await iterator.next()
  }

  let task2 = Task<Int?, Error> {
    var iterator = sut.makeAsyncIterator()
    return try await iterator.next()
  }

  // When: failing the channel
  sut.fail(Failure())

  // Then: the iterations are resumed with the error
  do {
    _ = try await (task1.value, task2.value)
  } catch {
    XCTAssertEqual(error as? Failure, Failure())
  }

  // When: requesting a next value
  var iterator = sut.makeAsyncIterator()
  let pastFailure = try await iterator.next()

  // Then: the past failure is nil
  XCTAssertNil(pastFailure)
}
```

FranzBusch replied:
Hm that's not really testing the same. I would like to see that test but with the same iterator calling next twice. Like this

```
  let task1 = Task<Int?, Error> {
    var iterator = sut.makeAsyncIterator()
    try await iterator.next() // Need to catch the values of both next and return them in an array to assert
    try await iterator.next()
  }
```

twittemb replied:
ok. 

Is that ok?

```
  func test_asyncThrowingChannel_resumes_consumer_when_fail_is_called() async throws {
    // Given: an AsyncThrowingChannel
    let sut = AsyncThrowingChannel<Int, Error>()

    // Given: suspended iteration
    let task = Task<Int?, Error> {
      var iterator = sut.makeAsyncIterator()

      do {
        _ = try await iterator.next()
      } catch {
        XCTAssertEqual(error as? Failure, Failure())
      }

      return try await iterator.next()
    }

    // When: failing the channel
    sut.fail(Failure())

    // Then: the iterations are resumed with the error and the next element is nil
    do {
      let collected = try await task.value
      XCTAssertNil(collected)
    } catch {
      XCTFail("The task should not fail, the past failure element should be nil, not a failure.")
    }
  }
```

FranzBusch replied:
I think almost, just one slight change
```
  func test_asyncThrowingChannel_resumes_consumer_when_fail_is_called() async throws {
    // Given: an AsyncThrowingChannel
    let sut = AsyncThrowingChannel<Int, Error>()

    // Given: suspended iteration
    let task = Task<Int?, Error> {
      var iterator = sut.makeAsyncIterator()

      do {
        _ = try await iterator.next()
        XCTFail("We expect the above call to throw")
      } catch {
        XCTAssertEqual(error as? Failure, Failure())
      }

      return try await iterator.next()
    }

    // When: failing the channel
    sut.fail(Failure())

    // Then: the iterations are resumed with the error and the next element is nil
    do {
      let collected = try await task.value
      XCTAssertNil(collected)
    } catch {
      XCTFail("The task should not fail, the past failure element should be nil, not a failure.")
    }
  }
````

Would be great if you could add that test

twittemb replied:
I had to do a few tweaks to make it pass ... nice catch.
I've pushed the updated version.

FranzBusch reviewed (changes requested):
Looks really good already. Left some more comments inline but we are getting close here! Thanks for all the work

twittemb commented:
> Looks really good already. Left some more comments inline but we are getting close here! Thanks for all the work

Thanks for the review. I'll address the comment later today or tomorrow.

phausler commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,149 @@
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+final class ChannelStorage<Element, Failure: Error>: @unchecked Sendable {
```
From what I can tell this technically does not need to be a reference type; the `stateMachine` and `ids` storage is relegated to `ManagedCriticalState` which handles the mutations.

This can be addressed as a follow-up change

twittemb replied:
done

phausler commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+  }
+
+  enum SendIsSuspendedAction {
+    case suspend
```
hmm putting a case of `none` infers the type should really just not have that case and be optional.

to me `.suspended` seems fine.

FranzBusch replied:
Yeah optional is fine as well. I really don't like suspended because we are already suspended here and nothing should really happen. We just need to wait for a consumer to show up.

twittemb replied:
it is now an optional

phausler commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+  private var state: State = .channeling(suspendedProducers: [], cancelledProducers: [], suspendedConsumers: [], cancelledConsumers: [])
+
+  enum SendAction {
+    case resumeConsumerAndExit(continuation: UnsafeContinuation<Element?, any Error>?)
```
Terminate is the parlance we have used in other documentation. But I agree that the proposed name Franz is offering is perhaps clearer.

twittemb replied:
renamed to `resumeConsumer`

phausler commented on Sources/AsyncAlgorithms/Channels/ChannelStorage.swift:
```diff
@@ -0,0 +1,149 @@
+
+  func generateId() -> Int {
+    self.ids.withCriticalRegion { ids in
+      defer { ids += 1 }
```
this should probably be `&+=`

```suggestion
      defer { ids &+= 1 }
```

twittemb replied:
done

phausler reviewed (approved):
Overall this looks pretty good; let me know when you are ready to land it.

twittemb commented:
> Overall this looks pretty good; let me know when you are ready to land it.

Thanks for the review, I think I'll push something tomorrow.

twittemb commented on Sources/AsyncAlgorithms/Channels/AsyncThrowingChannel.swift:
```diff
@@ -0,0 +1,62 @@
+  /// Sends an element to an awaiting iteration. This function will resume when the next call to `next()` is made
+  /// or when a call to `finish()` is made from another Task.
+  /// If the channel is already finished then this returns immediately
+  /// If the task is cancelled, this function will resume. Other sending operations from other tasks will remain active.
```
done.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+
+struct ChannelStateMachine<Element, Failure: Error>: Sendable {
+  private struct SuspendedProducer: Hashable {
+    let id: Int
```
done.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+    case resumeProducerAndConsumer(continuation: UnsafeContinuation<Element?, any Error>?)
+  }
+
+  mutating func sendIsSuspended(
```
renamed to `sendSuspended` to match the symmetrical `nextSuspended`. Also there is already a `SendAction` for the `send` function, it would have collided if both functions were named `send`.

If you fill that's not right I'll try to come up with something better.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+  mutating func sendIsSuspended(
+    continuation: UnsafeContinuation<Void, Never>,
+    element: Element,
+    producerId: Int
```
done.

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+        )
+        return .suspend
+
+      case .channeling(let suspendedProducers, var cancelledProducers, var suspendedConsumers, let cancelledConsumers):
```
done

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+    case resumeProducer(continuation: UnsafeContinuation<Void, Never>?)
+  }
+
+  mutating func sendIsCancelled(producerId: Int) -> SendIsCancelledAction {
```
renamed

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+    consumerId: Int
+  ) -> NextIsSuspendedAction {
+    switch self.state {
+      case .channeling(
```
done

twittemb commented on Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift:
```diff
@@ -0,0 +1,367 @@
+  }
+
+  enum NextIsSuspendedAction {
+    case suspend
```
it is now an optional

twittemb commented:
> Overall this looks pretty good; let me know when you are ready to land it.

I've addressed all the comments (minus one for the throwing behaviour).

In the end I can see a 20% increase in perfs.

FranzBusch commented on Tests/AsyncAlgorithmsTests/TestChannel.swift:
```diff
@@ -10,268 +10,167 @@
 
 import XCTest
-import AsyncAlgorithms
+@testable import AsyncAlgorithms
```
Do we still need the testable import? I think we are only using public API by now

twittemb replied:
nope. good catch. this is fixed.

FranzBusch reviewed (approved):
One last nit, but looks good to me now! Thanks for all the work!

twittemb commented:
> One last nit, but looks good to me now! Thanks for all the work!

Thanks a lot for your review.
